src/fifelse.c

```diff
@@ -50,10 +34,9 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b)
     UNPROTECT(2);
   }
 
-  // Check here the length of the different input variables.
   if (len1!=1 && len1!=len0) error("Length of 'yes' is %lld but must be 1 or length of 'test' (%lld).", len1, len0);
   if (len2!=1 && len2!=len0) error("Length of 'no' is %lld but must be 1 or length of 'test' (%lld).", len2, len0);
-  const int64_t amask = len1>1 ? INT64_MAX : 0;
+  const int64_t amask = len1>1 ? INT64_MAX : 0; // for scalar 'a' bitwise AND will reset iterator to first element: pa[i & amask] -> pa[0]
```

Comment:
oh i see now. helpful comment indeed!

and the thinking is the compiler will optimize away all the repeated `&` calls?

Reply:
bitwise AND of `any_number & 0` will result `0`, while of `any_number & max_number` will result `any_number`, due to the fact that `max_number` is represented by all `1`, i.e. `1 & 15` -> `0001 & 1111` results `0001`.